### PR TITLE
Ensure shape dimensions are within supported integer range (#566)

### PR DESCRIPTION
**Pull request by jmousseau**
## Proposed changes

In #566, the code below triggered an "unable to allocate" error because the shape's dimension overflowed `INT32_MAX`.

```python
mx.array(np.arange(1, 30 * 10**8))
```

Now a more specific error is thrown.

```python
>>> mx.array(np.arange(-1, 2147483647))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Shape dimension falls outside supported `int` range.

>>> mx.array([0] * 2147483648)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Shape dimension falls outside supported `int` range.
```

Note its still possible to produce shapes that trigger an allocation failure.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/array.h [hunk 1]
```diff
@@ -93,6 +95,21 @@ class array {
...
+    constexpr U min = static_cast<U>(std::numeric_limits<int>::min());
+    constexpr U max = static_cast<U>(std::numeric_limits<int>::max());
+
+    if ((is_signed && dim < min) || dim > max) {
+      throw std::invalid_argument(
+          "Shape dimension falls outside supported `int` range.");
+    }
+
+    return static_cast<int>(dim);
+  }
```
[the remark this answers is not in the document]
Comment (awni) [reply mlx/array.h:111]: Also I don't think we need this to be tepmlated (?). I would just have it take a `size_t` and simplify things a bit.

File: tests/array_tests.cpp [hunk 2]
```diff
@@ -69,6 +69,25 @@ TEST_CASE("test array basics") {
...
+  CHECK_EQ(array::check_shape_dim(static_cast<size_t>(dim_max)), dim_max);
+  CHECK_THROWS_AS(
+      array::check_shape_dim(static_cast<ssize_t>(dim_min) - 1),
+      std::invalid_argument);
+  CHECK_THROWS_AS(
+      array::check_shape_dim(static_cast<ssize_t>(dim_max) + 1),
+      std::invalid_argument);
+  CHECK_THROWS_AS(
+      array::check_shape_dim(static_cast<size_t>(dim_max) + 1),
+      std::invalid_argument);
```
Comment (awni) [inline tests/array_tests.cpp:89]: Maybe make this a python test and more integrated, something that tests e.g.:

```
>>> a = np.empty(2**32, np.bool_)
>>> mx.array(a).shape
(0,)
```

Comment (awni) [reply tests/array_tests.cpp:89]: That's only 4GB so it's not too bad I think..

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the contribution!! Left a couple comments, please address and then let me know when it's ready.

**Comment by awni** [comment]
@jmousseau are you coming back to this?

**Comment by jmousseau** [comment]
@awni Yes, I should have time to address comments this weekend. Apologies for the slowness/lack of response on my end.

File: mlx/array.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmousseau) [reply mlx/array.h:111]: Templated because `shape` returns `ssize_t` whereas `len` returns `size_t`. The reason I didn't place in `python/sr/array.cpp` was so that I could cover additional edge cases with C++ test cases.

File: tests/array_tests.cpp (hunk 2 above)
Comment (jmousseau) [reply tests/array_tests.cpp:89]: I've been burned by large allocations before, so excluded such tests. If we move `check_shape_dim` to `python/src/array.cpp`, will try adding python tests.

File: mlx/array.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (awni) [reply mlx/array.h:111]: Ok fair enough. I prefer it not be in `array.h`. You could put it in `mlx/utils.h`.

As for the template, I think you mean `shape` return `int`? It will be an always positive int, so you could cast it, but I don't feel too strongly about it.

[the remark this answers is not in the document]
Comment (jmousseau) [reply mlx/array.h]: Moved to `mlx/utils.h` and kept template as `py::shape` returns `ssize_t` and `py::len` returns `size_t`.

File: tests/array_tests.cpp (hunk 2 above)
Comment (jmousseau) [reply tests/array_tests.cpp]: Added python test case.

File: mlx/utils.cpp [hunk 3]
```diff
@@ -1,6 +1,8 @@
 // Copyright © 2023 Apple Inc.
 
+#include <limits>
 #include <sstream>
+#include <type_traits>
```
Comment (awni) [inline mlx/utils.cpp:5]: Do we need these added includes here?

Comment (jmousseau) [reply mlx/utils.cpp]: You're correct, these are unnecessary and have been removed.

**Review by awni — APPROVED** [review]
Looks great, thanks! I will merge when tests clear.
